### ISSUE TITLE
Type coercion

### ISSUE THREAD

**Issue by nitely** [mafintosh/is-my-json-valid#34]
Is there a way to coerce a given property?
I'm doing a custom validation the following way:

```javascript
var schema = {
  type: 'object',
  properties: {
    _id: {
      // type: ['string', 'mongoId'],
      format: 'mongoId'
    }
  }
};

var mongoId = function(_id) {
  // check if it's a string or a mongoId and validate it
};

var validate = validator(schema, {formats: {mongoId: mongoId}});
```

Now, I want to *auto coerce* the _id to an `OjectID` (mongo id type) whenever a string is received.
I'm thinking something along the lines of:

 ```javascript
var schema = {
  type: 'object',
  properties: {
    _id: {
      // type: ['string', 'mongoId'],
      format: 'mongoId',
      coerce: 'objectId',
    }
  }
};

var objectId = function(_id) {
  // coerce the _id to an objectId
  return coercedValue;
};

var validate = validator(schema, {coercers: {objectId: objectId}});

validate({_id: 'qwerty123'});
console.info(validate.data);  // {_id: ObjectID(qwerty123)}
```

The coercing would take place after all the other validations. My use case is specific to mongo but it'd be an useful feature for other type conversions.

**Comment by mafintosh** [comment]
Like a map function? @watson thoughts?

**Comment by nitely** [comment]
Yep. On a second thought, *coercion* is to specific. I'm thinking a way to manipulate a data property the way you want and return something different (or the same).

**Comment by filipedeschamps** [comment]
IMHO, this transformation logic should be inside your Model, not in the schema validator module.

**Comment by nitely** [comment]
@filipedeschamps my mongodb example may have been misleading. I still think the validation layer is the right place for data type coercion/manipulation.

**Comment by myndzi** [comment]
You use a validator when you have some (unknown) input data and you want some (known) output data. Superficially, this gives you validation as described: input data isn't valid, reject vs input data is valid, accept -- however, it's still desirable to accept coercible data (for example, with HTML form submissions, all the values become strings, even the numeric ones). In this case, the data is valid, but it is not in the format you have specified and know you want. Since the schema describes the format you want and the validator processes / iterates all the data and knows what needs to be changed, the validator is the ideal place to also perform coercion (since it must indeed coerce the data before deciding if it's valid). This module looks pretty sweet, coercion would make it even sweeter :)

**Comment by danieljuhl** [comment]
Does the JSON schema draft allow additional properties this way? I don't like that we would create invalid JSON schemas.

Wouldn't it be more correct to have a tool which could transform the JSON schema to a Mongoose model? Actually my team and I are working on such a tool, to convert from and to several JSON schema alike models/schemas/structures, which will soon be published to npm. It can convert a valid JSON schema to and from Mongoose (handling ObjectID and even Mongoose model references) and tcomb-form.

**Comment by myndzi** [comment]
The problem with that is now you're tied to a specific database driver/orm, or you have to write a bunch of them. It's better to do something generic that can be adapted to a specific use.

Re: defaults, I confused JSON Schema and Swagger. The 'default' property in a json schema is not what I thought it was.

**Comment by nitely** [comment]
@danieljuhl this issue is not database specific. Now I regret putting the mongo example... it was the first thing that came to mind. I do recognized it would use it that way.

Data cleaning it's something that's available in many validators (not so much in json validators since it's not part of the json schema specs).

**Comment by danieljuhl** [comment]
@nitely I don't have any concerns about creating a built-in data cleaning (like we already have the filter option). My concerns is purely on breaking the JSON schema draft/spec. I my opinion any such additional information, which can't be described in a valid JSON schema should be provided for the tool (whether its a standalone or built-in) as additional information instead of being part of the JSON schema.

**Comment by nitely** [comment]
@danieljuhl that's what I'm already doing. Something like this: `validate.coerce({myField: MyCoercerFunction, 'myField.myNestedField': myCoercerFunction})`

It feels way too awkward and you can't tell what's going on by just going through your schema.

I can't find anything about creating your own schema rules/properties/validations in the specs. Many other json validators do stuff like this, though.

**Comment by danieljuhl** [comment]
@nitely as far as I know, it is not possible, if the schema should still be a valid JSON schema (according to the draft). The only "valid" way to do it, would be to use the "title" or "description" meta fields, eg.
```
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "additionalProperties": false,
  "required": ["name"],
  "properties": {
    "_parent": {
      "description": "__coerce:objectId",
      "$ref": "#/definitions/objectid"
    },
    "name": {
      "type": "string"
    }
  },
  "definitions": {
    "objectid": {
      "type": "string",
      "pattern": "^[0-9a-f]{24}$"
    }
  }
}
```

Then we could have a coerce method like this:
```
validate.coerce(data, {
  objectId: coerceToObjectId
});
```

**Comment by watson** [comment]
I'm glad to hear that I'm not the only one having trouble interpreting JSON schema docs ;) Of cause if adding an extra property like suggested would make the schema invalid I vote for not doing it. This is a classic case of bundling an extra feature inside a simple module that already does one thing well. Especially someting that can easily be achieved by a separate module after the validation. That being said, it's intriguing to only have to describe your schema once and hence describe the coercion logic the same place. But my gut feeling currently says no :(

**Comment by myndzi** [comment]
I feel like coercion is worth it even without support for special/custom data types, especially given what is arguably one of the most useful cases for a module like this in node: validating form submissions. Even just making sure that numbers are numbers and strings are strings is enough to allow you to easily instantiate say, your mongo object: it will only be a one line assignment if you've already verified that you have correct data in the correct form.

**Comment by puzrin** [comment]
One more example - router data. Usually routers return strings, but very often params should be interpreted as integers. Situation is the same as with web forms - coerce data outside of validator is very inconvenient.

**Comment by RangerMauve** [comment]
Would it not make sense to coerce based on format or just default types?

For instance, if the input is a `String` but the type is a `number`, then try `parseFloat` on it.
If the type input is some custom format, it might make sense to have a function that coerces values of that format.

**Comment by cemtopkaya** [comment]
Is there any way to create an empty object from given schema?

**Comment by RangerMauve** [comment]
Yeah, I use [json-schema-defaults](https://github.com/chute/json-schema-defaults)

**Comment by cemtopkaya** [comment]
@RangerMauve thank you for your sharing. I want to create schemas in different files and create objects from their complex schemas. json-schema-defaults can do this? Or in node is there any other way to do this? Thanks in advance...

**Comment by ebdrup** [comment]
@cemtopkaya What do you want to use the object generated from the schema for?
If you want to use it documentation of what objects can be accepted by your schema, we use 
https://github.com/e-conomic/schemagic
and the `exampleJson` method.

**Comment by ebdrup** [comment]
@watson Additonal custom properties are allowed in JSON-schema:
Section 5.6 of core is easy to read: http://json-schema.org/latest/json-schema-core.html

```
5.6.  Extending JSON Schema

Implementations MAY choose to define additional keywords to JSON Schema. Save for explicit agreement, schema authors SHALL NOT expect these additional keywords to be supported by peer implementations. Implementations SHOULD ignore keywords they do not support.
```

**Comment by cemtopkaya** [comment]
@ebdrup Thank you soooo much for your quick answer. Now I am building node.js app and my needs are: 
- validate&delete props which are unneccesary before routing
<pre> app.post('/comments', f_validateAndDeleteRequestBody, function(req, res, next){ ...</pre>
- create objects from json schemas before response
<pre>app.post('/comments', f_validateAndDeleteRequestBody, function(req,res,next){
       ...
       res.json(xxxCreateObjectFromJsonSchema('./commentAdded.json'))
}</pre>

And I want to create separate json schema files like this:

<pre>- jsonSchemas/ 
    - def_address.json // address schema definition
    - person.json // ...properties:{ address: { $ref: 'def_address.json#address'} ....
    - ....
    - def_response.json // fixed response schema for every request but will be used in other schemas
    - commentAdded.json // properties:{ $ref: 'def_response.json#response' } </pre>

This way I'm planning to make every object map and when I want to change something that I will have to first change schemas. I hope I explained my goal very clear.

**Comment by RangerMauve** [comment]
@cemtopkaya The way I do that is to have a directory called 'schemas' somewhere in my source tree. Then you have alll your different types as JSON in there. As well I usually make an index.js which returns an object with all if the different schemas as properties. Then I can have something like
``` javascript
var Person = require("./schemas").Person
```

**Comment by iliocatallo** [comment]
no news on this side? It would be very handy for form validation!

**Comment by snesin** [comment]
I was able to accomplish this by adding a single line to the validator code.

At the bottom of **`is-my-json-valid/index.js`** add the following line:
`module.exports.types=types;`

The world is now your oyster! You can create custom types and coercers.

To manage the MongoDB's **`_id`** field, add the following to your project code:
```
// Step 1: Make Mongo's ObjectId class global.
GLOBAL.ObjectId = require('mongodb').ObjectID;

// Step 2: Include is-my-json-valid.
var jsonValidatorFactory = require('is-my-json-valid');

// Step 3: Create a validator for a custom type called ObjectId that also coerces.
// It must do this inside of an if statement, using a conditional operator.
// The logic is as follows:
//    Is the value a string of 24 hex characters?
//       YES: Change the property to an ObjectId and return it (truthy).
//       NO: Is the value a non-null object whos constructor is ObjectId?
//          YES: Return true.
//          NO: Return false.
jsonValidatorFactory.types.ObjectId = function(name) {
   return 'typeof '+name+' === "string" && /^[0-9a-f]{24}$/i.test('+name+') ? '+name+'=ObjectId('+name+') : '+name+' && '+name+'.constructor===ObjectId';
};
```
That is it! In the schema, you can use the type **`ObjectId`**, like so:
```
// Create a schema, using the new ObjectId type for the _id property.
var mySchema={
   required : true,
   type : "object",
   properties : {
      _id : {
         required : true,
         type : "ObjectId"
      }
   },
   additionalProperties : false
};

// Create a validator for the schema.
var myValidator=jsonValidatorFactory(mySchema);

// Create some test data with _id as a string.
var myData={ _id : "56b98ca35c55e4a8061a92d8" };
console.log("[START]  typeof _id:"+typeof(myData._id)+"   value:"+myData._id);

// Now call the validator with test data.
var isValid=myValidator(myData);
console.log("[FIRST] isValid:"+isValid+"  typeof _id:"+typeof(myData._id)+"   value:"+myData._id);
// isValid is true, and myData._id is now of type ObjectId! Woot!

// Test the validation against the data in its new form as well.
var isAlsoValid=myValidator(myData);
console.log("[SECOND] isValid:"+isValid+"  typeof _id:"+typeof(myData._id)+"   value:"+myData._id);
// isAlsoValid is also true!

// Try a bad _id string.
myData._id="56b9"; // Too short!
var isValid=myValidator(myData);
console.log("[THIRD] isValid:"+isValid+"  typeof _id:"+typeof(myData._id)+"   value:"+myData._id);
// isValid is false, and myData._id is still the unchanged string.
```
The console will show:
```
[START]  typeof _id:string   value:56b98ca35c55e4a8061a92d8
[FIRST] isValid:true  typeof _id:object   value:56b98ca35c55e4a8061a92d8
[SECOND] isValid:true  typeof _id:object   value:56b98ca35c55e4a8061a92d8
[THIRD] isValid:false  typeof _id:string   value:56b9
```
I will submit a pull request for the one line.
